packages/@ember/-internals/routing/lib/services/router.ts

```diff
@@ -474,6 +476,39 @@ export default class RouterService extends Service {
   */
 }
 
+if (EMBER_ROUTING_ROUTER_SERVICE_REFRESH) {
+  RouterService.reopen({
+    /**
+     * Refreshes all currently active routes, doing a full transition.
+     * If a route name is provided and refers to a currently active route,
+     * it will refresh only that route and its descendents.
+     * Returns a promise that will be resolved once the refresh is complete.
+     * All resetController, beforeModel, model, afterModel, redirect, and setupController
+     * hooks will be called again. You will get new data from the model hook.
+     *
+     * @method refresh
+     * @param {String} [pivotRouteName] the route to refresh (along with all child routes)
+     * @return Transition
+     * @public
+     */
+    refresh(pivotRouteName?: string) {
+      if (!pivotRouteName) {
+        return this._router._routerMicrolib.refresh();
+      }
+
+      assert(`The route "${pivotRouteName}" was not found`, this._router.hasRoute(pivotRouteName));
+      assert(
+        `The route "${pivotRouteName}" is currently not active`,
+        this.isActive(pivotRouteName)
+      );
+
+      let pivotRoute = this._router._routerMicrolib.getRoute(pivotRouteName) as Route;
```

Comment:
Why do we have to `.getRoute` here?

Reply:
Also, FWIW, `.getRoute` can return a promise (e.g. `ember-engines` or Embroider with lazy route bundles).

Reply:
> Why do we have to `.getRoute` here?

This comes back to [this earlier discussion](https://github.com/emberjs/ember.js/pull/19471#discussion_r597990897) where I questioned that as well. I did it like this because I had no idea how else to do it with the methods that are available on the router / microlib. And it seemed like it had logic related to engine routes as well. What's the proper way to do it exactly? Should I just look up the route in the container? Do I need to do anything specific for engine routes?
> Also, FWIW, .getRoute can return a promise (e.g. ember-engines or Embroider with lazy route bundles).

Yea, TS did indeed warn me about that :smile:. But the [`getRoute` implementation](https://github.com/emberjs/ember.js/blob/22bfcfdac0aeefcf333fb2d6697772934201b43b/packages/%40ember/-internals/routing/lib/system/router.ts#L190) in the Router had a `Route` return type, so I assumed it was ok.

Reply:
Ya, but ember-engines (and Embroider) override that method so that it sometimes returns a promise.

Reply:
Thanks for the info, that clears it up!

> What's the proper way to do it exactly? Should I just look up the route in the container? Do I need to do anything specific for engine routes?

I just retrieve the route instance through the container now. Let me know if I should do it differently.